### DISCUSSION TITLE
Using FirefoxPWA with LibreWolf?

### DISCUSSION THREAD

**Discussion by eNkKVS4HxHc9jaq** [filips123/PWAsForFirefox#106] [Ideas]
Are there any future plans to support LibreWolf (the private Firefox) instead of the native Firefox app? LibreWolf is basically just Firefox but without telemetry. I think it could be made to work with LibreWolf too since LibreWolf basically only changes the settings using the built in Enterprise out of the box experience for Firefox.

**Comment by filips123** [comment]
You can manually download LibreWolf and replace normal Firefox installation with it:

1. Download portable LibreWolf for your system. For Windows, portable ZIP version, and for Linux, portable `.tar.bz2`.
2. Replace existing Firefox installation made by FirefoxPWA with LibreWolf. On Windows it is at `C:\Program Files\FirefoxPWA\runtime` (but it may be different if you used non-default install path), and on Linux it is at `/usr/share/firefoxpwa/runtime`.
3. Symlink `firefox.exe`/`firefox` with `librewolf.exe`/`librewolf-bin`.

I quickly checked it on Windows and it seems to open sites correctly. Please let me know if you find any problems.

Automatic installation is currently not possible but I might add support for installing custom Firefox versions in the future.

**Comment by filips123** [comment]
@sh1ft-github I just released `0.5.0` which changes directory when runtime is installed. This means that you will need to manually move LibreWolf installation from old directory to the new one after you update to 0.5.0. This change will NOT affect your site or profile data. Extension will automatically open the default runtime installation page, but you will need to ignore it because it still only supports normal Firefox. This is hopefully the only time when runtime directory will change.

Old runtime locations:
-  Windows: `C:\Program Files\FirefoxPWA\runtime\` (depends on installation directory)
- Linux: `/usr/share/firefoxpwa/runtime/`
- macOS: `/usr/local/share/firefoxpwa/runtime/`

New runtime locations:
- Windows: `%APPDATA%\FirefoxPWA\runtime\`
- Linux: `$XDG_DATA_HOME/firefoxpwa/runtime/` or `$HOME/.local/share/firefoxpwa/runtime/`
- macOS: `$HOME/Library/Application Support/firefoxpwa/runtime/`

**Comment by dajix350** [comment]
Could you make a video showing how to do this?

**Comment by filips123** [comment]
I currently cannot create a video, but here are more detailed installation/update steps:

**Windows:**

1. Download [latest `.win64.zip` file from LibreWolf releases](https://gitlab.com/librewolf-community/browser/windows/-/releases).
2. Create empty directory `%APPDATA%\FirefoxPWA\runtime\` (or remove its content if it exists).
3. Extract downloaded file and copy **content** (without "root" directory) of `librewolf-VERSION\LibreWolf\` to `%APPDATA%\FirefoxPWA\runtime\`.
4. Open Command Line in that directory and run `mklink firefox.exe librewolf.exe`.

**Linux:**

1. Download [latest `.x86_64.tar.bz2` file from LibreWolf releases](https://gitlab.com/librewolf-community/browser/linux/-/releases).
2. Create empty directory `$HOME/.local/share/firefoxpwa/runtime/` (or remove its content if it exists).
3. Extract downloaded file and copy its **content** (without "root" directory) to `$HOME/.local/share/firefoxpwa/runtime/`.
4. Open Terminal in that directory and run: `ln librewolf firefox`.

**macOS:**

1. Download [latest `.dmg` file from LibreWolf releases](https://gitlab.com/librewolf-community/browser/macos/-/releases).
2. Create empty directory `$HOME/Library/Application Support/firefoxpwa/runtime/Firefox.app/` (or remove its content if it exists).
3. Extract downloaded file and copy **content** (without "root" directory) of `LibreWolf\LibreWolf.app\` to `$HOME/Library/Application Support/firefoxpwa/runtime/Firefox.app/`.
4. Open Terminal in `$HOME/Library/Application Support/firefoxpwa/runtime/Firefox.app/Contents/MacOS/` and run: `ln librewolf firefox`.

If this doesn't work, please check [the log files](https://github.com/filips123/FirefoxPWA/wiki/Troubleshooting).

**Comment by dajix350** [comment]
I tried, but on the Setup FirefoxPWA page it said 

Error: The program has been detected and is already installed, but it has incompatible version that does not work correctly with the current version of the extension. You need to reinstall the program in order to make this extension work correctly. 

I installed the app via Npacked GUI on win 10. firefoxpwa.log was empty

Screenshot if it helps: 
![Screenshot 2021-09-06 143057](https://user-images.githubusercontent.com/74509183/132160874-5dffddd0-cb6c-4676-aebf-8357957b52a2.png)

**Comment by filips123** [comment]
I think this happened because I recently released 1.1.0, but Npacked hasn't updated the package yet. You can wait until Npacked updates FirefoxPWA package or manually install update using MSI installer.

**Comment by filips123** [comment]
I've added additional installation instructions in 1.3.0 that explain how to use alternative Firefox-based browsers. I hope this will be enough to let users know how to use LibreWolf if they want.

![screenshot](https://user-images.githubusercontent.com/16626308/147392632-a72591c2-b7fa-40e6-9d94-9b539c9427e1.png)

Fully supporting LibreWolf or alternative browsers as an "official" browser (e.g. auto install) would probably need too much work, so I'll just close this issue for now and move it to discussions. I may change this decision in the future or accept PRs if anyone creates them.

**Comment by michaelBelsanti** [comment]
Has anybody had luck with this? I haven't gotten past the connector installation yet using Librewolf. I have the arch package of both firefox-pwa and librewolf and have followed the instructions above, but haven't made any progress.

**Reply by filips123** [reply]
Are you stuck on a step where you have to install the connector or when you have to install the runtime?

If the former, this is caused by LibreWolf [having different native manifests location](https://librewolf.net/docs/faq/#how-do-i-get-native-messaging-to-work) than the normal Firefox. You should be able to fix this by running:

```bash
ln -s ~/.mozilla/native-messaging-hosts ~/.librewolf/native-messaging-hosts
sudo ln -s /usr/lib/mozilla/native-messaging-hosts /usr/lib/librewolf/native-messaging-hosts
```

If the latter, you need to symlink/copy `/usr/lib/librewolf` to `$XDG_DATA_HOME/firefoxpwa/runtime` (if `$XDG_DATA_HOME` is set) or `$HOME/.local/share/firefoxpwa/runtime` (if it is not), and symlink `librewolf` binary as `firefox` inside that directory.

**Reply by michaelBelsanti** [reply]
It was the connector and that fixed it for me! Thanks a ton! Are there any plans to add that step to the documentation? (assuming it's not already there, I might've just missed it.)

**Reply by filips123** [reply]
Yeah, I will add it. I didn't knew before LibreWolf has its own manifests location.

**Reply by asumbek** [reply]
About the symlinking LibreWolf binary as `firefox` in the runtime directory, for me (in EndavourOS) it doesn't work. But when I tried copying the contents of `/usr/lib/librewolf` into the runtime directory and then symlink `librewolf` as `firefox`, it works.

**Comment by candrapersada** [comment]
how to update LibreWolf or other Firefox based browser to latest version in PWA?

**Reply by filips123** [reply]
I think LibreWolf does not have own auto-updater, so will have to manually download its latest version and copy it to the runtime location, similarly to when you installed it.

**Comment by candrapersada** [comment]
why PWAsForFirefox doesn't work with just symlink firefox.exe from librewolf.exe?
`%APPDATA%\FirefoxPWA\runtime\`
![gambar](https://github.com/user-attachments/assets/690ec6b5-6ae3-45ab-909f-793b625b8441)

**Reply by filips123** [reply]
Because all other files that Firefox/LibreWolf needs also need to be in that directory. So, you will have to copy all of them.

**Comment by moritztim** [comment]
Any plans on making this work out of the box? Allowing the user to specify which runtime to use like
```
firefoxpwa runtime install librewolf
```

**Reply by filips123** [reply]
Not for now. There are too many Firefox forks to support them officially. Also, LibreWolf appearently does not have built-in auto updates, so you would still have to manually keep it up-to-date.

**Reply by moritztim** [reply]
>  There are too many Firefox forks to support them officially.

Fair enough, yeah
> LibreWolf appearently does not have built-in auto updates, so you would still have to manually keep it up-to-date.

Not if I have a package manager

**Reply by filips123** [reply]
> Not if I have a package manager

Yeah, but in this case, command to download LibreWolf from their website won't be useful as would download normal without auto-updates. Once #660 is merged and finished, you could probably use it to link runtime to your system LibreWolf version, or you xan already manually symlink or copy it.

**Comment by bgmeulem** [comment]
Interestingly, if I simply symlink the runtime directory:
```sh
ln -s /usr/lib/librewolf/* ~/.local/share/firefoxpwa/runtime
```
(and of course also `ln -s librewolf firefox`)
I get permission denied. If i copy (instead of symlink) the following directories and files
```sh
application.ini
_autoconfig.cfg
browser/
defaults/
dependlibs.list
librewolf.cfg
omni.ja
platform.ini
precomplete
```
It still does not work, and opening webapps simply opens librewolf with a default profile (distinct from both the webapp profile and the user profile).
Copying the entire directory
```sh
cp -r /usr/lib/librewolf/* ~/.local/share/firefoxpwa/runtime
```
Does work. I'm curious to which files it needs write access?

**Reply by filips123** [reply]
It needs to write `_autoconfig.cfg` and `defaults/pref/autoconfig.js` in the runtime directory.
